text/0000-jsbox.md

```diff
@@ -0,0 +1,341 @@
+
+* Many `neon` methods require `&mut Context`. While a `JsBox` is borrowed, most `neon` methods could not be called. We could potentially mitigate this by performing a sweeping audit and relaxing constraints where a shared reference is acceptable
+* Multiple `JsBox`, even if they are unrelated, could not be borrowed at the same time because there is no way to statically prove they do not point to the same underlying data
+* Impairs ergonomics on APIs that do not require mutable borrowing since `Deref` cannot be implemented
+
+### `JsCell`
+
+Neon could provide a `JsCell` type that is identical to `JsBox`, but also wraps the underlying `T` in a `std::cell::RefCell`. This allows the user to also mutably borrow the value.
+
+```rust
+impl<T> JsCell where T: Send + 'static {
+    /// This should only fail if the VM is in a throwing state or is
+    /// in the process of stopping.
+    pub fn new<'a, C: Context<'a>>(
+        cx: &mut C,
+        v: T,
+    ) -> NeonResult<JsCell>;
+    /// May panic if already mutably borrowed
+    pub fn borrow<'a, C: Context<'a>>(
+        &self,
+        cx: &mut C,
+    ) -> Ref<'a, T>;
+    pub fn try_borrow<'a, C: Context<'a>>(
+        &self,
+        cx: &mut C,
+    ) -> Result<Ref<'a, T>, BorrowError>;
+    /// May panic if already borrowed
+    pub fn borrow_mut<'a, C: Context<'a>>(
+        &self,
+        cx: &mut C,
+    ) -> RefMut<'a, T>;
+    pub fn try_borrow_mut<'a, C: Context<'a>>(
+        &self,
+        cx: &mut C,
+    ) -> Result<RefMut<'a, T>, BorrowMutError>;
+    /// The most common `RefCell` methods are provided on `JsCell` as a
+    /// convenience. Other methods may can be accessed by calling `as_cell`
+    pub fn as_cell<'a, C: Context<'a>>(
+        &self,
+        cx: &mut C,
+    ) -> &RefCell<T>;
+}
+```
+
+This approach has several significant drawbacks:
+
+* Increases the API substantial
+* Adds an abstraction that not all users might need
+* Prevents implementing `Deref`
+
+Instead, Neon can lean on documentation to guide users towards `RefCell` if they need interior mutability.
+
+### Naming
+
+Several other names were considered:
+
+* `JsExternal`. Matches the name in N-API documentation (external reference); however, `cx.external(v)` might be confusing.
+* `JsRef`. Also, matches N-API but, does not match Rust naming conventions.
+* `JsCell`. Only appropriate for the `RefCell` approach.
+
+Overall, `JsBox<T>` behaves very similar to a `std::boxed::Box<T>`.
+
+### Leverage existing borrow API
+
+An existing [`neon::borrow`](https://docs.rs/neon/0.4.0/neon/borrow/index.html) implementation exists. However, the current implementation has multiple soundness holes:
+
+* `Buffer` can be mutably borrowed multiple times
+* `Lock` does not prevent overlapping memory regions
+* Multiple [`Lock`](https://docs.rs/neon/0.4.0/neon/context/struct.Lock.html) can be created allowing aliased mutable borrows
```

Comment:
I'm still kind of shocked about this soundness hole. I could've sworn I checked for it and panic.

Conceptually, I still find the idea of being able to use RAII and Rust mutability rules to lock the main JS thread to be pretty compelling. But I like how you've at least decoupled that idea from the idea of safely exposing Rust data owned by the JS engine back to Rust.

I wonder if we might still have a place for the `Lock` API just as a way to have self-documenting critical sections. OTOH, it might end up being a cute idea that's not particularly useful anymore. Curious what you think?

Reply:
I don't think the `Lock` API fits here because nothing is being locked. I don't understand what you mean by RAII to lock the main js thread. The existing API doesn't take any action to lock the the VM.

Reply:
@dherman I thought about this some more and it finally clicked. I think the `Lock` guard API only makes sense for accessing data from `Buffer`.

When accessing a `Buffer`, Neon needs to uphold Rusts guarantees on ownership and mutability of the slice. However, since the data is not owned by Rust, there may exist references unknown to Rust.

For example, imagine a scenario where a user borrows a `&[u8]` to the `Buffer`. According to Rust invariants, this data will not change during the lifetime of the borrow. However, if JavaScript is invoked, e.g. by calling a `JsFunction`, the data could be mutated, violating the invariant.

The `Lock` API solves this by holding an exclusive reference to `Context`. The `Lock` statically prevents calling any Neon methods that could potentially mutate the slice.

This problem does not exist for `JsBox`. The data is owned by Rust and _can't_ be mutated by pure JavaScript. At best, the user could invoke JavaScript that calls another Neon method. However, that Neon method would still be bound by Rusts dynamic borrow checking (`RefCell`). It's simpler, more ergonomic, and no less safe to return shared references without a Lock/Guard on `JsBox`.

Lastly, there _is_ an edge case where the `Lock` API might not uphold invariants. If the `Buffer` was created as an external buffer (memory created and managed externally), it's possible for other non-Neon code to mutate these types of buffers. Neon could use `n-api` methods for detecting externals and refuse to lend them. Additionally, neon could provide `unsafe fn borrow_unchecked` methods for reading data from externals.

Reply:
It took me a couple of reads but I think you've really nailed the difference:

- Rust data (`JsBox`) **can** be protected by Rust's `Cell` invariants, so we **don't** need to worry about reborrows via VM re-entrancy;
- C/C++ data (`JsArrayBuffer`) **can't** be protected by Rust's `Cell` invariants, so we **do** need to worry about reborrows via VM re-entrancy.

So if I understand you, you're saying the `Lock` API makes sense to keep, but specifically and only for typed arrays.

On the one hand, it's nice that we can borrow a `JsBox` more ergonomically. On the other, the asymmetry is a little frustrating. It's a shame there's no way (AFAIK) to dynamically borrow an `ArrayBuffer`, which we could then use to make this safe without having to lock the JS VM. You can permanently detach its backing buffer, but that doesn't help this use case; we'd need a temporary, reversible version of detaching.

> The `Lock` statically prevents calling any Neon methods that could potentially mutate the slice.

Exactly! This is definitely subtle (if ingenious—thank Niko for the idea), and is probably worth some better explanation in docs, assuming we keep the API.

If we only end up needing it for `ArrayBuffer` and not `JsBox`, I wonder if we want a less general name than `Borrow`, especially since `JsBox` doesn't use it. Maybe something that makes more explicit that it's a stop-the-world borrow, so the connection to `Lock` is clearer?

> Lastly, there is an edge case where the `Lock` API might not uphold invariants. If the `Buffer` was created as an external buffer (memory created and managed externally), it's possible for other non-Neon native code to mutate these types of buffers. Neon could use n-api methods for detecting externals and refuse to lend them. Additionally, neon could provide `unsafe fn borrow_unchecked` methods for reading data from externals.

Good catch. That makes sense to me.

So, some of this stuff is out of scope for this RFC, but probably worth summarizing the issues in the "open questions" section so we can capture them for subsequent work.

Reply:
@dherman I like the clear naming of `borrow` for the API. However, we could move  it from `neon::borrow` to `neon::binary` to clarify the relationship to `JsArrayBuffer`.